File: LibLiteLog/src/main/java/io/github/tomgarden/lib/lite_log/Logger.java
package io.github.tomgarden.lib.lite_log;

import io.github.tomgarden.lib.lite_log.function.Function2;

/**
 * describe : 为了在 AOSP 中阅读代码写的工具类 , 刚开始用 kotlin 后来发现项目环境不支持 , 所以 java 重写一次
 * <p>
 * author : Create by tom , on 2020/9/16-3:04 PM
 * github : https://github.com/TomGarden
 */
public class Logger {
    static final int VERBOSE = 2;
    static final int DEBUG = 3;
    static final int INFO = 4;
    static final int WARN = 5;
    static final int ERROR = 6;
    static final int ASSERT = 7;

    Printer printer = new LogPrinter();

    public static Logger INSTANCE = getInstance();

    private Logger() {
    }

    private static Logger getInstance() {
        if (INSTANCE == null) {
            synchronized (Logger.class) {
                if (INSTANCE == null) {
                    INSTANCE = new Logger();
                }
            }
        }

        return INSTANCE;
    }

    public Logger clearLogStrategies() {
        printer.clearLogStrategies();
        return this;
    }

    //region default logcat
    public LogStrategy getDefLogcatStrategy() {
        return printer.getDefLogcatStrategy();
    }

    public Logger setDefLogcatStrategy(LogStrategy defLogcatStrategy) {
        printer.setDefLogcatStrategy(defLogcatStrategy);
        return this;
    }

    public Logger defLogcatMethodCount(int methodCount) {
        setMethodCount(printer.getDefLogcatStrategy(), methodCount);
        return this;
    }

    public Logger defLogcatMethodOffset(int methodOffset) {
        setMethodOffset(printer.getDefLogcatStrategy(), methodOffset);
        return this;
    }

    public Logger defLogcatShowThreadInfo(boolean showThreadInfo) {
        setShowThreadInfo(printer.getDefLogcatStrategy(), showThreadInfo);
        return this;
    }

    public Logger defLogcatTag(String tag) {
        setTag(printer.getDefLogcatStrategy(), tag);
        return this;
    }

    public Logger defLogcatIsLoggable(Function2<Integer, String, Boolean> isLoggable) {
        setIsLoggable(printer.getDefLogcatStrategy(), isLoggable);
        return this;
    }
    //endregion default logcat

    //region temporary logcat
    public LogStrategy getTempLogcatStrategy() {
        return printer.getTemporaryLogcatStrategy();
    }

    public Logger setTempLogcatStrategy(LogStrategy temporaryLogcatStrategy) {
        printer.setTemporaryLogcatStrategy(temporaryLogcatStrategy);
        return this;
    }

    public Logger tempLogcatMethodCount(int methodCount) {
        setMethodCount(printer.unNullTemporaryLogcatStrategy(), methodCount);
        return this;
    }

    public Logger tempLogcatMethodOffset(int methodOffset) {
        setMethodOffset(printer.unNullTemporaryLogcatStrategy(), methodOffset);
        return this;
    }

    public Logger tempLogcatShowThreadInfo(boolean showThreadInfo) {
        setShowThreadInfo(printer.unNullTemporaryLogcatStrategy(), showThreadInfo);
        return this;
    }

    public Logger tempLogcatTag(String tag) {
        setTag(printer.unNullTemporaryLogcatStrategy(), tag);
        return this;
    }

    public Logger tempLogcatIsLoggable(Function2<Integer, String, Boolean> isLoggable) {
        setIsLoggable(printer.unNullTemporaryLogcatStrategy(), isLoggable);
        return this;
    }

    public Logger tempJustMsg(int methodCount) {
        this.tempLogcatMethodCount(methodCount)
                .tempLogcatShowThreadInfo(false);
        return this;
    }

    public Logger tempJustMsg() {
        int methodOffset = printer.unNullTemporaryLogcatStrategy().methodCount;
        tempJustMsg(methodOffset);
        return this;
    }
    //endregion temporary logcat

    //region default disk
    public LogStrategy getDefDiskStrategy() {
        return printer.getDefDiskStrategy();
    }

    public DiskLogTxtStrategy getDefDiskLogTxtStrategy() {
        LogStrategy strategy = printer.getDefDiskStrategy();
        if (strategy == null) return null;
        if (strategy instanceof DiskLogTxtStrategy) {
            return (DiskLogTxtStrategy) strategy;
        }
        return null;
    }

    /*check disk log file path*/
    public String getDefDiskStrategyLogFilePath() {
        LogStrategy strategy = printer.getDefDiskStrategy();
        if (strategy == null) return null;
        if (strategy instanceof DiskLogTxtStrategy) {
            return ((DiskLogTxtStrategy) strategy).getLogFilePath().invoke();
        }
        return null;
    }

    public Logger setDefDiskStrategy(LogStrategy defDiskStrategy) {
        printer.setDefDiskStrategy(defDiskStrategy);
        return this;
    }

    public Logger defDiskMethodCount(int methodCount) {
        setMethodCount(printer.getDefDiskStrategy(), methodCount);
        return this;
    }

    public Logger defDiskMethodOffset(int methodOffset) {
        setMethodOffset(printer.getDefDiskStrategy(), methodOffset);
        return this;
    }

    public Logger defDiskShowThreadInfo(boolean showThreadInfo) {
        setShowThreadInfo(printer.getDefDiskStrategy(), showThreadInfo);
        return this;
    }

    public Logger defDiskTag(String tag) {
        setTag(printer.getDefDiskStrategy(), tag);
        return this;
    }

    public Logger defDiskIsLoggable(Function2<Integer, String, Boolean> isLoggable) {
        setIsLoggable(printer.getDefDiskStrategy(), isLoggable);
        return this;
    }
    //endregion default disk

    //region temporary disk
    public LogStrategy getTempDiskStrategy() {
        return printer.getTemporaryDiskStrategy();
    }

    public Logger setTempDiskStrategy(LogStrategy temporaryDiskStrategy) {
        printer.setTemporaryDiskStrategy(temporaryDiskStrategy);
        return this;
    }

    public Logger tempDiskMethodCount(int methodCount) {
        setMethodCount(printer.unNullTemporaryDiskStrategy(), methodCount);
        return this;
    }

    public Logger tempDiskMethodOffset(int methodOffset) {

        setMethodOffset(printer.unNullTemporaryDiskStrategy(), methodOffset);
        return this;
    }

    public Logger tempDiskShowThreadInfo(boolean showThreadInfo) {

        setShowThreadInfo(printer.unNullTemporaryDiskStrategy(), showThreadInfo);
        return this;
    }

    public Logger tempDiskTag(String tag) {

        setTag(printer.unNullTemporaryDiskStrategy(), tag);
        return this;
    }

    public Logger tempDiskIsLoggable(Function2<Integer, String, Boolean> isLoggable) {

        setIsLoggable(printer.unNullTemporaryDiskStrategy(), isLoggable);
        return this;
    }
    //endregion temporary disk


    //region set logStrategy
    private Logger setMethodCount(LogStrategy logStrategy, int methodCount) {
        if (logStrategy == null) {
            nullPointLog();
        } else {
            logStrategy.methodCount = methodCount;
        }
        return this;
    }

    private Logger setMethodOffset(LogStrategy logStrategy, int methodOffset) {
        if (logStrategy == null) {
            nullPointLog();
        } else {
            logStrategy.methodOffset = methodOffset;
        }
        return this;
    }

    private Logger setShowThreadInfo(LogStrategy logStrategy, boolean showThreadInfo) {
        if (logStrategy == null) {
            nullPointLog();
        } else {
            logStrategy.showThreadInfo = showThreadInfo;
        }
        return this;
    }

    private Logger setTag(LogStrategy logStrategy, String tag) {
        if (logStrategy == null) {
            nullPointLog();
        } else {
            logStrategy.tag = tag;
        }
        return this;
    }

    private Logger setIsLoggable(LogStrategy logStrategy, Function2<Integer, String, Boolean> isLoggable) {
        if (logStrategy == null) {
            nullPointLog();
        } else {
            logStrategy.isLoggable = isLoggable;
        }
        return this;
    }

    private Logger nullPointLog() {
        LogStrategy logStrategy = getTempLogcatStrategy();
        if (logStrategy != null) {
            logStrategy.log(ERROR, "Null point exception, 'Logger.setXxx(..)' is failed");
        }
        return this;
    }
    //endregion set logStrategy


    /**
     * General log function that accepts all configurations as parameter
     */
    public Logger log(int priority, String message, Throwable throwable, boolean withSingleFile) {
        printer.log(priority, message, throwable, withSingleFile);
        return this;
    }

    public Logger d(String message, Object... args) {
        printer.d(message, args);
        return this;
    }

    public Logger d(Object any) {
        printer.d(any);
        return this;
    }

    public Logger e(String message, Object... args) {
        printer.e(false, message, args);
        return this;
    }

    public Logger e(boolean withSingleFile, String message, Object... args) {
        printer.e(withSingleFile, message, args);
        return this;
    }

    public Logger e(Throwable throwable, String message, Object... args) {
        printer.e(throwable, false, message, args);
        return this;
    }

    public Logger e(Throwable throwable, boolean withSingleFile, String message, Object... args) {
        printer.e(throwable, withSingleFile, message, args);
        return this;
    }

    public Logger i(String message, Object... args) {
        printer.i(message, args);
        return this;
    }

    public Logger v(String message, Object... args) {
        printer.v(message, args);
        return this;
    }

    public Logger w(String message, Object... args) {
        printer.w(message, args);
        return this;
    }

    /**
     * Tip: Use this for exceptional situations to log
     * ie: Unexpected errors etc
     */
    public Logger wtf(String message, Object... args) {
        printer.wtf(message, args);
        return this;
    }
}

File: LibLog/src/main/java/io/github/tomgarden/lib/log/Logger.kt
package io.github.tomgarden.lib.log

import java.io.File

/**
 * <pre>
 * ┌────────────────────────────────────────────
 * │ LOGGER
 * ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
 * │ Standard logging mechanism
 * ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
 * │ But more pretty, simple and powerful
 * └────────────────────────────────────────────
</pre> *
 *
 * <h3>How to use it</h3>
 * Initialize it first
 * <pre>`
 * Logger.addLogAdapter(new AndroidLogAdapter());
`</pre> *
 *
 * And use the appropriate static Logger methods.
 *
 * <pre>`
 * Logger.d("debug");
 * Logger.e("error");
 * Logger.w("warning");
 * Logger.v("verbose");
 * Logger.i("information");
 * Logger.wtf("What a Terrible Failure");
`</pre> *
 *
 * <h3>String format arguments are supported</h3>
 * <pre>`
 * Logger.d("hello %s", "world");
`</pre> *
 *
 * <h3>Collections are support ed(only available for debug logs)</h3>
 * <pre>`
 * Logger.d(MAP);
 * Logger.d(SET);
 * Logger.d(LIST);
 * Logger.d(ARRAY);
`</pre> *
 *
 * <h3>Json and Xml support (output will be in debug level)</h3>
 * <pre>`
 * Logger.json(JSON_CONTENT);
 * Logger.xml(XML_CONTENT);
`</pre> *
 *
 * <h3>Customize Logger</h3>
 * Based on your needs, you can change the following settings:
 *
 *  * Different [LogStrategy]
 *  * Different [LogPrinter]
 *
 *
 * @see LogStrategy
 */
object Logger {

    val VERBOSE = 2
    val DEBUG = 3
    val INFO = 4
    val WARN = 5
    val ERROR = 6
    val ASSERT = 7

    var printer: Printer = LogPrinter()

    fun clearLogStrategies() = printer.clearLogStrategies()

    //region default disk
    fun getDefDiskStrategy(): LogStrategy? = printer.defDiskStrategy

    fun getDefDiskLogTxtStrategy(): DiskLogTxtStrategy? =
        printer.defDiskStrategy as DiskLogTxtStrategy?

    /*check disk log file path*/
    fun getDefDiskStrategyLogFilePath(): String? {
        val defDiskStrategy = printer.defDiskStrategy as DiskLogTxtStrategy?
        val path =
            defDiskStrategy?.let { defDiskStrategy ->
                defDiskStrategy.logFilePath.invoke()
            } ?: "Not setDefDiskStrategy"

        return path
    }

    fun setDefDiskStrategy(defDiskStrategy: LogStrategy?) {
        printer.defDiskStrategy = defDiskStrategy
    }

    fun defDiskMethodCount(methodCount: Int): Logger =
        setMethodCount(printer.defDiskStrategy, methodCount)

    fun defDiskMethodOffset(methodOffset: Int): Logger =
        setMethodOffset(printer.defDiskStrategy, methodOffset)

    fun defDiskShowThreadInfo(showThreadInfo: Boolean): Logger =
        setShowThreadInfo(printer.defDiskStrategy, showThreadInfo)

    fun defDiskTag(tag: String): Logger = setTag(printer.defDiskStrategy, tag)

    fun defDiskIsLoggable(isLoggable: ((priority: Int, tag: String) -> Boolean)): Logger =
        setIsLoggable(printer.defDiskStrategy, isLoggable)
    //endregion default disk

    //region temporary disk
    fun getTempDiskStrategy(): LogStrategy? = printer.temporaryDiskStrategy

    fun setTempDiskStrategy(temporaryDiskStrategy: LogStrategy?) {
        printer.temporaryDiskStrategy = temporaryDiskStrategy
    }

    fun tempDiskMethodCount(methodCount: Int): Logger =
        setMethodCount(printer.unNullTemporaryDiskStrategy(), methodCount)

    fun tempDiskMethodOffset(methodOffset: Int): Logger =
        setMethodOffset(printer.unNullTemporaryDiskStrategy(), methodOffset)

    fun tempDiskShowThreadInfo(showThreadInfo: Boolean): Logger =
        setShowThreadInfo(printer.unNullTemporaryDiskStrategy(), showThreadInfo)

    fun tempDiskTag(tag: String): Logger = setTag(printer.unNullTemporaryDiskStrategy(), tag)

    fun tempDiskIsLoggable(isLoggable: ((priority: Int, tag: String) -> Boolean)): Logger =
        setIsLoggable(printer.unNullTemporaryDiskStrategy(), isLoggable)
    //endregion temporary disk

    //region default logcat
    fun getDefLogcatStrategy(): LogStrategy? = printer.defLogcatStrategy

    fun setDefLogcatStrategy(defLogcatStrategy: LogStrategy?): Logger {
        printer.defLogcatStrategy = defLogcatStrategy
        return this
    }

    fun defLogcatMethodCount(methodCount: Int): Logger =
        setMethodCount(printer.defLogcatStrategy, methodCount)

    fun defLogcatMethodOffset(methodOffset: Int): Logger =
        setMethodOffset(printer.defLogcatStrategy, methodOffset)

    fun defLogcatShowThreadInfo(showThreadInfo: Boolean): Logger =
        setShowThreadInfo(printer.defLogcatStrategy, showThreadInfo)

    fun defLogcatTag(tag: String): Logger = setTag(printer.defLogcatStrategy, tag)

    fun defLogcatIsLoggable(isLoggable: ((priority: Int, tag: String) -> Boolean)): Logger =
        setIsLoggable(printer.defLogcatStrategy, isLoggable)
    //endregion default logcat

    //region temporary logcat
    fun getTempLogcatStrategy(): LogStrategy? = printer.temporaryLogcatStrategy

    fun setTempLogcatStrategy(temporaryLogcatStrategy: LogStrategy?): Logger {
        printer.temporaryLogcatStrategy = temporaryLogcatStrategy
        return this
    }

    fun tempLogcatMethodCount(methodCount: Int): Logger =
        setMethodCount(printer.unNullTemporaryLogcatStrategy(), methodCount)

    fun tempLogcatMethodOffset(methodOffset: Int): Logger =
        setMethodOffset(printer.unNullTemporaryLogcatStrategy(), methodOffset)

    fun tempLogcatShowThreadInfo(showThreadInfo: Boolean): Logger =
        setShowThreadInfo(printer.unNullTemporaryLogcatStrategy(), showThreadInfo)

    fun tempLogcatTag(tag: String): Logger =
        setTag(printer.unNullTemporaryLogcatStrategy(), tag)

    fun tempLogcatIsLoggable(isLoggable: ((priority: Int, tag: String) -> Boolean)): Logger =
        setIsLoggable(printer.unNullTemporaryLogcatStrategy(), isLoggable)

    fun tempJustMsg(): Logger {
        val methodCount = printer.unNullTemporaryLogcatStrategy().methodCount
        return tempJustMsg(methodCount)
    }

    fun tempJustMsg(methodCount: Int): Logger {
        return tempLogcatMethodCount(methodCount).tempLogcatShowThreadInfo(false)
    }
    //endregion temporary logcat


    //region set logStrategy
    private fun setMethodCount(logStrategy: LogStrategy?, methodCount: Int): Logger {
        logStrategy?.methodCount = methodCount
        logStrategy ?: nullPointLog()
        return this
    }

    private fun setMethodOffset(logStrategy: LogStrategy?, methodOffset: Int): Logger {
        logStrategy?.methodOffset = methodOffset
        logStrategy ?: nullPointLog()
        return this
    }

    private fun setShowThreadInfo(logStrategy: LogStrategy?, showThreadInfo: Boolean): Logger {
        logStrategy?.showThreadInfo = showThreadInfo
        logStrategy ?: nullPointLog()
        return this
    }

    private fun setTag(logStrategy: LogStrategy?, tag: String): Logger {
        logStrategy?.tag = tag
        logStrategy ?: nullPointLog()
        return this
    }

    private fun setIsLoggable(
        logStrategy: LogStrategy?,
        isLoggable: ((priority: Int, tag: String) -> Boolean)
    ): Logger {
        logStrategy?.isLoggable = isLoggable
        logStrategy ?: nullPointLog()
        return this
    }

    private fun nullPointLog() {
        getTempLogcatStrategy()?.log(
            ERROR,
            "Null point exception, 'Logger.setXxx(..)' is failed"
        )
    }
    //endregion set logStrategy


    /**
     * General log function that accepts all configurations as parameter
     */
    fun log(priority: Int, message: String?, throwable: Throwable?, withSingleFile: Boolean) =
        printer.log(priority, message, throwable, withSingleFile)

    fun d(message: String?, vararg args: Any) = printer.d(message, *args)

    fun d(any: Any?) = printer.d(any)

    fun e(message: String?, vararg args: Any) = printer.e(false, message, *args)

    fun e(withSingleFile: Boolean, message: String?, vararg args: Any) =
        printer.e(withSingleFile, message, args)

    fun e(throwable: Throwable?, message: String, vararg args: Any) =
        printer.e(throwable, false, message, *args)

    fun e(throwable: Throwable?, withSingleFile: Boolean, message: String, vararg args: Any) =
        printer.e(throwable, withSingleFile, message, *args)

    fun e(any: Any?) = printer.e(any)

    fun i(message: String?, vararg args: Any) = printer.i(message, *args)

    fun i(any: Any?) = printer.i(any)

    fun v(message: String?, vararg args: Any) = printer.v(message, *args)

    fun v(any: Any?) = printer.v(any)

    fun w(message: String?, vararg args: Any) = printer.w(message, *args)

    fun w(any: Any?) = printer.w(any)

    /**
     * Tip: Use this for exceptional situations to log
     * ie: Unexpected errors etc
     */
    fun wtf(message: String?, vararg args: Any) = printer.wtf(message, *args)

    fun assert(assert: Boolean, message: String?, vararg args: Any) =
        printer.wtf(assert, message, *args)

    fun wtf(any: Any?) = printer.wtf(any)

    fun assert(assert: Boolean, any: Any?) = printer.wtf(assert, any)

    /**
     * Formats the given json content and print it
     */
    fun json(json: String?) = printer.json(json)

    /**
     * Formats the given xml content and print it
     */
    fun xml(xml: String?) = printer.xml(xml)


    //***************************************************************************************
    //                                 文件读取
    //***************************************************************************************

    fun readFile(filePath: String): String = Utils.readFile(filePath)

    fun getLogFiles(fileName: String, fileExtend: String): MutableList<File> {
        val emptyList: MutableList<File> = mutableListOf()

        val logStrategy: LogStrategy = getDefDiskStrategy() ?: return emptyList

        var folderPath: String? = null
        if (logStrategy is DiskLogTxtStrategy) {
            folderPath = logStrategy.logFilePath.invoke()
        } else if (logStrategy is DiskLogTxtStrategy) {
            folderPath = logStrategy.logFilePath.invoke()
        }

        if (folderPath.isNullOrEmpty()) {
            return emptyList
        } else {
            return Utils.getLogFiles(folderPath, fileName, fileExtend)
        }

    }


    fun getCrashLogFiles(): MutableList<File> =
        getLogFiles(WriteHandler.crashLogName, WriteHandler.fileExtend)

    fun getNormalLogFiles(): MutableList<File> =
        getLogFiles(WriteHandler.normalLogName, WriteHandler.fileExtend)

    fun clearLogFiles() {
        getNormalLogFiles().forEach { logFile -> logFile.delete() }
        getCrashLogFiles().forEach { logFile -> logFile.delete() }
    }
}
